Fix: Can no longer slice csr_matrix in 1.3.0

ananyashreyjain opened the pull request:
Fixes
#11200

This PR will provide a certain degree of leniency in array types used for storing boolean indices when converting them to integers. For example, after this PR, CSR matrices can be sliced directly with boolean indices even if the boolean indices are not stored as `numpy.ndarray`.

tylerjereddy commented:
Ah, I see the sparse failures are actually *separate* from this sparse change, possibly due to NumPy 1.18: https://github.com/scipy/scipy/issues/11266

Confusing!

ananyashreyjain commented:
@tylerjereddy, even I think that the failures are separate from this change, originating due to Numpy 1.18.0 #11270 and #11266

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -276,9 +276,9 @@ def _unpack_index(index):
             'except boolean indexing where matrix and index '
             'are equal shapes.')
-    if isinstance(row, np.ndarray) and row.dtype.kind == 'b':
+    if hasattr(row, 'ndim') and row.dtype.kind == 'b':
```
Leave a comment here to explain why we're checking for `ndim` despite not accessing it. Or better, define a small helper function `_is_boolean_index` that performs these checks.

ananyashreyjain replied:
Making a separate helper function will be a good idea.

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -325,4 +325,4 @@ def _check_ellipsis(index):
     if idx.ndim > 1:
         raise IndexError('invalid index shape')
-    return idx.nonzero()[0]
+    return np.where(idx != 0)[0]
```
This should be equivalent to `np.where(idx)[0]`.

ananyashreyjain commented:
Instead of proceeding on basis on presence of attribute `ndim` which may not cover all kinds of array types (for eg, list). I think it will be better to convert to `numpy.ndarray` (if possible) and then proceed. This will cover all kinds of array types that can be used for slicing.

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -322,6 +324,20 @@ def _check_ellipsis(index):
+def _compatible_boolean_index(idx):
+    """Returns a boolean index array that can be converted to
+    integer array. Returns None if no such array exists """
+    if hasattr(idx, '__iter__'):
```
This seems a bit too permissive. For instance, a dictionary with boolean keys would be an accepted index.

How about:

```python
if not hasattr(idx, 'ndim') or not hasattr(idx, 'dtype'):
    idx = np.asanyarray(idx)
if idx.dtype.kind == 'b':
    return idx
return None
```

ananyashreyjain replied:
Yes, this seems to simpler and better. I will make the changes.

perimosocordiae commented:
@ananyashreyjain could you add a test case that indexes a sparse matrix with a list of booleans, and checks that it's equivalent to the same thing using an array of booleans?

ananyashreyjain commented:
@perimosocordiae I have added some checks to avoid the performance hit for the case with non-boolean element types. I have also added a test case to ensure that indexing with boolean list works fine. The failure in Travis CI doesn't look related to this PR.

ananyashreyjain commented:
Okay, this won't work with 2D Index list. I will change it back.

ananyashreyjain commented:
Okay, this won't work with 2D array types of boolean indices. Some more modifications will be needed to make it work for 2D array types.

ananyashreyjain commented:
This should now be compatible with all array types single or double dimensional without any performance hit for the case with non-boolean element types.

ananyashreyjain commented on scipy/sparse/_index.py:
```diff
@@ -266,8 +266,13 @@ def _unpack_index(index):
+            row, col = index, slice(None)
+        elif idx.ndim < 2:
+            return _boolean_index_to_array(idx), slice(None)
+        elif idx.ndim == 2:
```
Earlier `col` was set to `None` for non `numpy.ndarray`s  irrespective of the dimension which is why the 2D index array couldn't have worked.

ananyashreyjain commented on scipy/sparse/_index.py:
```diff
@@ -266,8 +266,13 @@ def _unpack_index(index):
         else:
             raise IndexError('invalid number of indices')
     else:
-        row, col = index, slice(None)
```
Earlier `col` was set to `None` for non `numpy.ndarray`  irrespective of the dimension, which is why the 2D index array couldn't have worked.

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -322,7 +329,33 @@ def _check_ellipsis(index):
+                        return idx
+    elif idx.dtype.kind == 'b':
+        return idx
+    return None
```
This is a lot of nesting as written. How about:

```python
def _maybe_bool_ndarray(idx):
    idx = np.asanyarray(idx)
    if idx.dtype.kind == 'b':
        return idx
    return None

def _first_element_bool(idx):
    try:
        first = next(iter(idx), None)
    except TypeError:
        return None
    if isinstance(first, bool):
        return True
    return _first_element_bool(first)

def _compatible_boolean_index(idx):
    if hasattr(idx, 'ndim') or _first_element_bool(idx):
        return _maybe_bool_ndarray(idx)
    return None
```

ananyashreyjain replied:
Thanks, I have made the changes.

ananyashreyjain commented:
@perimosocordiae Thanks for all your help. I have the made the style changes and I hope this PR can be merged now.

perimosocordiae commented:
Merged. Thanks for sticking with this, @ananyashreyjain!
